{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ab0258c3-c161-4916-a92b-052b7e20fd67",
   "metadata": {},
   "source": [
    "## Imports and Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2c5ed3f-5ebd-4327-aed2-3dabd1e5535b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import itertools as it\n",
    "import os\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import random\n",
    "import time\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a90f1f2-47c9-4883-b0a4-8e6f0617dab5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# variables\n",
    "closepairs = [['C', 'T'], ['A', 'T'], ['A', 'G']] # needed for ``close'' error mode"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a642dc-14b7-4fe5-8a82-26d4f6bc77d8",
   "metadata": {},
   "source": [
    "## Dictionaries and Frequencies\n",
    "This need not be changed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fca773a7-d23b-4195-ba4b-2a8fb1507fe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_dict(S):\n",
    "    d = { }\n",
    "    for s in S:\n",
    "        if not s in d:\n",
    "            d[s] = 0\n",
    "    return d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4152a04d-49b9-40cf-b8ee-17b064065438",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define character sets\n",
    "othchars = {'N', '\\n'} # Expected characters that should be ignored\n",
    "nucchars = {'T', 'A', 'C', 'G'} # Characters that should be read as part of the RNA # , 'B', 'K', 'M', 'S', 'R', 'W', 'Y'\n",
    "corenucchars = ['T', 'A', 'C', 'G'] # Subset of characters of RNA that are valid\n",
    "cdnchars = [] # All triplets of nucchars\n",
    "for c in it.product(nucchars, repeat=3):\n",
    "    codon = c[0]+c[1]+c[2]\n",
    "    cdnchars.append(codon)\n",
    "corecdnchars = [] # All triplets of corenucchars\n",
    "for c in it.product(corenucchars, repeat=3):\n",
    "    codon = c[0]+c[1]+c[2]\n",
    "    corecdnchars.append(codon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ea1f5b6-af59-423b-950a-d2200ac7b946",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Codon dictionary\n",
    "codondict = {\n",
    "    \"TTT\": \"Phenyl-alanine\",\n",
    "    \"TTC\": \"Phenyl-alanine\",\n",
    "    \"TTA\": \"Leucine\",\n",
    "    \"TTG\": \"Leucine\",\n",
    "    \"TCT\": \"Serine\",\n",
    "    \"TCC\": \"Serine\",\n",
    "    \"TCA\": \"Serine\",\n",
    "    \"TCG\": \"Serine\",\n",
    "    \"TAT\": \"Tyrosine\",\n",
    "    \"TAC\": \"Tyrosine\",\n",
    "    \"TAA\": \"Stop\",\n",
    "    \"TAG\": \"Stop\",\n",
    "    \"TGT\": \"Cysteine\",\n",
    "    \"TGC\": \"Cysteine\",\n",
    "    \"TGA\": \"Stop\",\n",
    "    \"TGG\": \"Tryptophan\",\n",
    "    \"CTT\": \"Leucine\",\n",
    "    \"CTC\": \"Leucine\",\n",
    "    \"CTA\": \"Leucine\",\n",
    "    \"CTG\": \"Leucine\",\n",
    "    \"CCT\": \"Proline\",\n",
    "    \"CCC\": \"Proline\",\n",
    "    \"CCA\": \"Proline\",\n",
    "    \"CCG\": \"Proline\",\n",
    "    \"CAT\": \"Histidine\",\n",
    "    \"CAC\": \"Histidine\",\n",
    "    \"CAA\": \"Glutamine\",\n",
    "    \"CAG\": \"Glutamine\",\n",
    "    \"CGT\": \"Arginine\",\n",
    "    \"CGC\": \"Arginine\",\n",
    "    \"CGA\": \"Arginine\",\n",
    "    \"CGG\": \"Arginine\",\n",
    "    \"ATT\": \"Isoleucine\",\n",
    "    \"ATC\": \"Isoleucine\",\n",
    "    \"ATA\": \"Isoleucine\",\n",
    "    \"ATG\": \"Methionine\",\n",
    "    \"ACT\": \"Threonine\",\n",
    "    \"ACC\": \"Threonine\",\n",
    "    \"ACA\": \"Threonine\",\n",
    "    \"ACG\": \"Threonine\",\n",
    "    \"AAT\": \"Asparagine\",\n",
    "    \"AAC\": \"Asparagine\",\n",
    "    \"AAA\": \"Lysine\",\n",
    "    \"AAG\": \"Lysine\",\n",
    "    \"AGT\": \"Serine\",\n",
    "    \"AGC\": \"Serine\",\n",
    "    \"AGA\": \"Arginine\",\n",
    "    \"AGG\": \"Arginine\",\n",
    "    \"GTT\": \"Valine\",\n",
    "    \"GTC\": \"Valine\",\n",
    "    \"GTA\": \"Valine\",\n",
    "    \"GTG\": \"Valine\",\n",
    "    \"GCT\": \"Alanine\",\n",
    "    \"GCC\": \"Alanine\",\n",
    "    \"GCA\": \"Alanine\",\n",
    "    \"GCG\": \"Alanine\",\n",
    "    \"GAT\": \"Aspartic Acid\",\n",
    "    \"GAC\": \"Aspartic Acid\",\n",
    "    \"GAA\": \"Glutamic Acid\",\n",
    "    \"GAG\": \"Glutamic Acid\",\n",
    "    \"GGT\": \"Glycine\",\n",
    "    \"GGC\": \"Glycine\",\n",
    "    \"GGA\": \"Glycine\",\n",
    "    \"GGG\": \"Glycine\",\n",
    "}\n",
    "aminonames = []\n",
    "for c in corecdnchars:\n",
    "    if not codondict[c] in aminonames:\n",
    "        aminonames.append(codondict[c])\n",
    "\n",
    "rev_codondict = make_dict(aminonames) # dictionary from amino acid to codons\n",
    "for c in corecdnchars:\n",
    "    if  rev_codondict[codondict[c]] == 0:\n",
    "        rev_codondict[codondict[c]] = []\n",
    "    rev_codondict[codondict[c]].append(c)\n",
    "        \n",
    "print(aminonames)\n",
    "print(rev_codondict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12426da1-a513-4db3-8509-e151ca760dcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load distribution of codons\n",
    "distdict = {'TTT': 5379930, 'TTC': 3489330, 'TTG': 3346091, 'TTA': 2854192, 'TCT': 3713851, 'TCC': 3333211, 'TCG': 827472, 'TCA': 3560263, 'TGT': 3671777, 'TGC': 3358600, 'TGG': 4198572, 'TGA': 3896165, 'TAT': 2717421, 'TAC': 1971858, 'TAG': 1765414, 'TAA': 2772671, 'CTT': 3604102, 'CTC': 3418446, 'CTG': 4886439, 'CTA': 2028492, 'CCT': 3994829, 'CCC': 3584074, 'CCG': 1395220, 'CCA': 4126496, 'CGT': 850978, 'CGC': 1167168, 'CGG': 1393193, 'CGA': 935301, 'CAT': 3177732, 'CAC': 3021415, 'CAG': 4927410, 'CAA': 3494785, 'GTT': 2540730, 'GTC': 2064226, 'GTG': 3312170, 'GTA': 1790976, 'GCT': 3339168, 'GCC': 3427017, 'GCG': 1193087, 'GCA': 3324824, 'GGT': 2258681, 'GGC': 3176324, 'GGG': 3156393, 'GGA': 3967764, 'GAT': 2672919, 'GAC': 2444849, 'GAG': 3952791, 'GAA': 4332881, 'ATT': 3552064, 'ATC': 2458843, 'ATG': 3518926, 'ATA': 2584899, 'ACT': 2896694, 'ACC': 2744211, 'ACG': 915076, 'ACA': 3584179, 'AGT': 2959982, 'AGC': 3579694, 'AGG': 3790603, 'AGA': 4636759, 'AAT': 3574871, 'AAC': 2726839, 'AAG': 4257699, 'AAA': 5772129}\n",
    "\n",
    "# count amino acids, and their percentage\n",
    "total = 0\n",
    "da = make_dict(aminonames)\n",
    "dt = make_dict(aminonames)\n",
    "\n",
    "for c in corecdnchars:\n",
    "    da[codondict[c]] += distdict[c]\n",
    "    total += distdict[c]\n",
    "    \n",
    "for c in aminonames:\n",
    "    dt[c] = da[c] / total\n",
    "\n",
    "# sort and graph\n",
    "X = list(aminonames)\n",
    "X.sort(reverse=True, key=lambda x : da[x])\n",
    "Y = [da[x] for x in X]\n",
    "    \n",
    "fig, ax = plt.subplots(figsize =(16, 9))\n",
    "plt.bar(X,Y,width=0.5,color=[\"tab:blue\", \"tab:orange\", \"tab:green\", \"tab:red\"])\n",
    "plt.title(\"Amino Frequency\")\n",
    "plt.xlabel('Aminoacid')\n",
    "plt.ylabel('Frequency')\n",
    "plt.xticks(rotation=45)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6118835-aa10-47f2-9192-85254079c0e5",
   "metadata": {},
   "source": [
    "## General Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe42c894-c78a-4149-ae49-8c03565054c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# empty for now\n",
    "# like - really general. Helper functions for error modes should be grouped at the respective error mode cell"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "730d5632-61ac-4aed-b9d9-9bb091314142",
   "metadata": {},
   "source": [
    "## Generating the RNA Sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "157a4e12-c404-4209-8c3b-eca61e72bc44",
   "metadata": {},
   "outputs": [],
   "source": [
    "# returns a codon given a number between 0 and 1\n",
    "def find_codon(p):\n",
    "    current = 0\n",
    "    curprob = 0        \n",
    "    \n",
    "    while current < len(aminonames):\n",
    "        curprob += dt[aminonames[current]]\n",
    "        if p < curprob:\n",
    "            return aminonames[current]\n",
    "        current += 1\n",
    "        \n",
    "    return aminonames[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "946b9f25-4a37-4ea7-b3f1-4db1b4bffbeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# return whether there is an error given a number between 0 and 1\n",
    "def has_error(p, p_err):\n",
    "    if p < p_err:\n",
    "        return 1\n",
    "    return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc0585f2-3bb0-4cc7-b03d-cee55d42925f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generates input with an RNA sequence of n codons, and accompanying error, modified RNA, and error code \n",
    "def gen_RNAerr(n, p_err, errmode):\n",
    "    rna = \"\"\n",
    "    err = []\n",
    "    mod_rna = \"\"\n",
    "    err_code = []\n",
    "    \n",
    "    # create a sequence with n codons\n",
    "    for i in range(n):\n",
    "        r = random.uniform(0, 1)\n",
    "        cdn = random.choice(rev_codondict[find_codon(r)])\n",
    "        rna += cdn\n",
    "    \n",
    "    # for each nucleotide, determine whether there is an error, and compute the resulting nucleotide and error code by the error mode\n",
    "    for i in range(3*n):\n",
    "        r = random.uniform(0, 1)\n",
    "        cur_err = has_error(r, p_err)\n",
    "        err.append(cur_err)\n",
    "        \n",
    "        mod_char, err_bit = errmode[0](rna[i],cur_err)\n",
    "        err_code.append(err_bit)\n",
    "        mod_rna += mod_char\n",
    "\n",
    "    return rna, err, mod_rna, err_code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dcebfda-21ff-450c-9cd2-7b178f1402f6",
   "metadata": {},
   "source": [
    "## Reading the RNA Sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51c02afa-1a28-4f7a-87f9-d118efce91a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# determines all codons we get when error-correcting codon c with error code e using error mode errmode\n",
    "def pos_codons(c, e, errmode):\n",
    "    p = [errmode[1](c[i], e[i]) for i in range(3)]\n",
    "    return [x + y + z for x in p[0] for y in p[1] for z in p[2]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcad7b1e-8f75-466d-a4dd-fdcdcb247fac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# finds whether an rna sequence can be read successfully.\n",
    "# we interpret reading successfully as: for each codon, either a codon representing the original amino acid is recorded OR ALL possible codons after correction of the recorded codon represent the original amino acid\n",
    "def can_read(rna, err, mod_rna, err_code, errmode):\n",
    "    n = len(rna)\n",
    "    if not n == len(mod_rna):\n",
    "        print(\"Invalid length\")\n",
    "    \n",
    "    output = True\n",
    "    nsucc_err = 0 # number of characters that can be individually corrected successfully\n",
    "    ntot_err = 0\n",
    "    nsucc = [0 for i in range(4)] # nsucc[i] = number of successful reads with i errors in codon\n",
    "    ntot = [0 for i in range(4)] # nsucc[i] = number of codons with i errors\n",
    "    \n",
    "    for i in range(math.floor(n/3)):\n",
    "        # get information for the current codon\n",
    "        real_codon = rna[3*i : 3*i + 3]\n",
    "        mod_codon = mod_rna[3*i : 3*i + 3]\n",
    "        err_codon = err[3*i : 3*i + 3]\n",
    "        err_code_codon = err_code[3*i : 3*i + 3]\n",
    "        \n",
    "        \n",
    "        # compute success for errors in current codon\n",
    "        for i in range(3):\n",
    "            if err_codon[i] == 0:\n",
    "                continue\n",
    "            ntot_err += 1\n",
    "            if real_codon[i] == mod_codon[i]:\n",
    "                nsucc_err += 1\n",
    "                continue\n",
    "            corr_codons = errmode[1](mod_codon[i], err_code_codon[i])\n",
    "            \n",
    "            valid = True\n",
    "            for c in corr_codons:\n",
    "                if not c == real_codon[i]:\n",
    "                    valid = False\n",
    "                    break\n",
    "            \n",
    "            if valid:\n",
    "                nsucc_err += 1\n",
    "        \n",
    "        # compute number of errors in current codon\n",
    "        n_err = 0\n",
    "        for e in err_codon:\n",
    "            if e == 1:\n",
    "                n_err += 1\n",
    "        \n",
    "        ntot[n_err] += 1\n",
    "        \n",
    "        ident_codons = rev_codondict[codondict[real_codon]] # find all codons representing same aminoacid as real codon\n",
    "        \n",
    "        if mod_codon in ident_codons: # the output of the machine corresponds to the same codon\n",
    "            nsucc[n_err] += 1\n",
    "            continue\n",
    "        \n",
    "        # check if all possible codons after error correction are correct\n",
    "        errcorr_codons = pos_codons(mod_codon, err_code_codon, errmode)\n",
    "        valid = True\n",
    "        for c in errcorr_codons:\n",
    "            if not c in ident_codons:\n",
    "                valid = False\n",
    "                break\n",
    "        \n",
    "        if valid:\n",
    "            nsucc[n_err] += 1\n",
    "            continue\n",
    "        \n",
    "        output = False\n",
    "    \n",
    "    return output, nsucc_err, nsucc, ntot_err, ntot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39782cd9-7d18-4852-84c6-fbe40e7d0f5d",
   "metadata": {},
   "source": [
    "## Doing Experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e81d10a3-9740-4ef8-90a3-67c998df8d62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# does a single experiment on a number of trials\n",
    "def do_experiment(codoncount, p_err, expcount, errmode):\n",
    "    successes = 0\n",
    "    succ_err_read = 0\n",
    "    tot_err_read = 0\n",
    "    succ_read = [0 for i in range(4)]\n",
    "    tot_read = [0 for i in range(4)]\n",
    "    \n",
    "    for i in range(expcount):\n",
    "        rna, err, mod_rna, err_code = gen_RNAerr(codoncount, p_err, errmode)\n",
    "        read, succ_err, succ, tot_err, tot = can_read(rna, err, mod_rna, err_code, errmode)\n",
    "        if read:\n",
    "            successes += 1\n",
    "            \n",
    "        succ_err_read += succ_err\n",
    "        tot_err_read += tot_err\n",
    "        \n",
    "        for i in range(4):\n",
    "            succ_read[i] += succ[i]\n",
    "            tot_read[i] += tot[i]\n",
    "        \n",
    "    return successes, succ_err_read, succ_read, tot_err_read, tot_read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2946be08-cbcc-4368-b499-88a402055a83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# does an experiment for all specified sequence lenghts (in codons), and error probabilities\n",
    "def do_experiments(codoncounts, p_errs, expcounts, errmode):\n",
    "    out = [[ 0 for j in p_errs ] for i in codoncounts]\n",
    "    \n",
    "    for i in range(len(codoncounts)):\n",
    "        for j in range(len(p_errs)):\n",
    "            res = do_experiment(codoncounts[i], p_errs[j], expcounts, errmode)\n",
    "            out[i][j] = res\n",
    "    \n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf606d62-69e3-41d6-adb6-fc095adc317e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# returns several figues:\n",
    "# - heatmap\n",
    "# - combined line plot with all fixed lengths \n",
    "# - combined line plot with all fixed p_errs\n",
    "# - list of plots, one for each fixed length, in order\n",
    "# - list of plots, one for each fixed p_err, in order\n",
    "# parameters are as input and output from the experiments\n",
    "# pass an empty string as title to have the title as only the statistic used\n",
    "# statistic chooses the heatmap to create:\n",
    "# =0 : probability of success\n",
    "# =1 : probability correct reading for error\n",
    "# =2 : probability correct codon no errors\n",
    "# =3 : probability correct codon one error\n",
    "# =4 : probability correct codon two errors\n",
    "# =5 : probability correct codon three errors\n",
    "# =6 : probability correct codon any errors\n",
    "def drawfigures_experiments(codoncounts, p_errs, expcount, data_in, statistic=0, title=\"Placeholder\", show = True, save=False):\n",
    "    #print(\"Title:\", title)\n",
    "    #print(\"Figures shown:\", show)\n",
    "    #print(\"Figures saved:\", save)\n",
    "    \n",
    "    wbar = min([p_errs[i + 1] - p_errs[i] for i in range(len(p_errs)) if i+1 < len(p_errs)])\n",
    "    \n",
    "    if statistic < 0 or statistic > 6: # if incorrectly specified, set to default\n",
    "        statistic = 0\n",
    "    \n",
    "    title_full = title\n",
    "    \n",
    "    if not title_full == \"\":\n",
    "        title_full += \": \"\n",
    "        \n",
    "    if statistic == 0:\n",
    "        title_full += \"P_Succ\"\n",
    "    if statistic == 1:\n",
    "        title_full += \"P_SuccCor Single Error\"\n",
    "    if statistic == 2:\n",
    "        title_full += \"P_SuccCor Codon No Errors\"\n",
    "    if statistic == 3:\n",
    "        title_full += \"P_SuccCor Codon One Error\"\n",
    "    if statistic == 4:\n",
    "        title_full += \"P_SuccCor Codon Two Errors\"\n",
    "    if statistic == 5:\n",
    "        title_full += \"P_SuccCor Codon Three Errors\"\n",
    "    if statistic == 6:\n",
    "        title_full += \"P_SuccCor Codon Any Errors\"\n",
    "    \n",
    "    data = np.zeros((len(codoncounts), len(p_errs)))\n",
    "    for i in range(len(codoncounts)):\n",
    "        for j in range(len(p_errs)):\n",
    "            if statistic == 0: # probability success\n",
    "                if expcount == 0:\n",
    "                    data[i][j] = 1\n",
    "                    continue\n",
    "                data[i][j] = data_in[i][j][0] / expcount\n",
    "            if statistic == 1: # probability correct reading for error\n",
    "                if data_in[i][j][3] == 0:\n",
    "                    data[i][j] = 1\n",
    "                    continue\n",
    "                data[i][j] = data_in[i][j][1] / data_in[i][j][3]\n",
    "            if statistic == 2: # probability correct codon no errors\n",
    "                if data_in[i][j][4][0] == 0:\n",
    "                    data[i][j] = 1\n",
    "                    continue\n",
    "                data[i][j] = data_in[i][j][2][0] / data_in[i][j][4][0]\n",
    "            if statistic == 3: # probability correct codon one error\n",
    "                if data_in[i][j][4][1] == 0:\n",
    "                    data[i][j] = 1\n",
    "                    continue\n",
    "                data[i][j] = data_in[i][j][2][1] / data_in[i][j][4][1]\n",
    "            if statistic == 4: # probability correct codon two errors\n",
    "                if data_in[i][j][4][2] == 0:\n",
    "                    data[i][j] = 1\n",
    "                    continue\n",
    "                data[i][j] = data_in[i][j][2][2] / data_in[i][j][4][2]\n",
    "            if statistic == 5: # probability correct codon three errors\n",
    "                if data_in[i][j][4][3] == 0:\n",
    "                    data[i][j] = 1\n",
    "                    continue\n",
    "                data[i][j] = data_in[i][j][2][3] / data_in[i][j][4][3]\n",
    "            if statistic == 6: # probability correct codon any errors\n",
    "                if data_in[i][j][4][0] + data_in[i][j][4][1] + data_in[i][j][4][2] + data_in[i][j][4][3] == 0:\n",
    "                    data[i][j] = 1\n",
    "                    continue\n",
    "                data[i][j] = (data_in[i][j][2][0] + data_in[i][j][2][1] + data_in[i][j][2][2] + data_in[i][j][2][3]) / (data_in[i][j][4][0] + data_in[i][j][4][1] + data_in[i][j][4][2] + data_in[i][j][4][3])\n",
    "                \n",
    "    figcnt = 0\n",
    "    figheat = plt.figure(figcnt, figsize=(7.5, 5.625))\n",
    "    ax = figheat.add_axes(sns.heatmap(data, xticklabels=p_errs, yticklabels=codoncounts, linewidth=1, annot=True, vmin=0, vmax=1))\n",
    "    plt.title(title_full)\n",
    "    plt.xlabel('p_err')\n",
    "    plt.ylabel('#Codons')\n",
    "    if save: plt.savefig(\"output/heatfig/\" + str(statistic) + \"_heat_\" + title + \".pdf\")\n",
    "    if show: plt.show()\n",
    "    plt.close()\n",
    "    figcnt += 1\n",
    "    \n",
    "    figcodonline = plt.figure(figcnt, figsize=(7.5, 5.625))\n",
    "    ax = plt.axes()\n",
    "    for i in range(len(codoncounts)):\n",
    "        ax.plot(p_errs, data[i], label=str(codoncounts[i]))\n",
    "    plt.legend(title = \"#Codons\", loc=\"upper right\")\n",
    "    plt.title(title_full)\n",
    "    plt.xlabel('p_err')\n",
    "    plt.ylabel('Probability')\n",
    "    plt.xlim([0, 1])\n",
    "    if save: plt.savefig(\"output/codonfig/\" + str(statistic) + \"_line_combined_\" + title + \".pdf\")\n",
    "    if show: plt.show()\n",
    "    plt.close()\n",
    "    figcnt += 1\n",
    "    \n",
    "    figperrline = plt.figure(figcnt, figsize=(7.5, 5.625))\n",
    "    ax = plt.axes()\n",
    "    for i in range(len(p_errs)):\n",
    "        ax.plot(codoncounts, data[:,i], label=str(p_errs[i]))\n",
    "    plt.legend(title = \"p_err\", loc=\"upper right\")\n",
    "    plt.title(title_full)\n",
    "    plt.xlabel('#Codons')\n",
    "    plt.ylabel('Probability')\n",
    "    plt.xlim([0, codoncounts[-1] + 1])\n",
    "    if save: plt.savefig(\"output/perrfig/\" + str(statistic) + \"_line_combined_\" + title + \".pdf\")\n",
    "    if show: plt.show()\n",
    "    plt.close()\n",
    "    figcnt += 1\n",
    "    \n",
    "    figcods = [[-1, -1] for i in range(len(codoncounts))]\n",
    "    figerrs = [[-1, -1] for i in range(len(p_errs))]\n",
    "\n",
    "    for i in range(len(codoncounts)):\n",
    "        fig = plt.figure(figcnt, figsize=(7.5, 5.625))\n",
    "        ax = plt.axes()\n",
    "        ax.plot(p_errs, data[i])\n",
    "        plt.title(title_full + \", #Codons: \" + str(codoncounts[i]))\n",
    "        plt.xlabel('p_err')\n",
    "        plt.ylabel('Probability')\n",
    "        plt.xlim([0, 1])\n",
    "        figcods[i][0] = fig\n",
    "        if save: plt.savefig(\"output/codonfig/\" + str(statistic) + \"_line_codons\" + str(codoncounts[i]) + \"_\" + title + \".pdf\")\n",
    "        if show: plt.show()\n",
    "        plt.close()\n",
    "        figcnt += 1\n",
    "        \n",
    "        fig = plt.figure(figcnt, figsize=(7.5, 5.625))\n",
    "        ax = plt.axes()\n",
    "        ax.bar(p_errs, data[i],width=wbar, color=[\"tab:blue\", \"tab:orange\", \"tab:green\", \"tab:red\"])\n",
    "        plt.title(title_full + \", #Codons: \" + str(codoncounts[i]))\n",
    "        plt.xlabel('p_err')\n",
    "        plt.ylabel('Probability')\n",
    "        plt.xlim([0, 1])\n",
    "        figcods[i][1] = fig\n",
    "        if save: plt.savefig(\"output/codonfig/\" + str(statistic) + \"_bar_codons\" + str(codoncounts[i]) + \"_\" + title + \".pdf\")\n",
    "        if show: plt.show()\n",
    "        plt.close()\n",
    "        figcnt += 1\n",
    "\n",
    "    for i in range(len(p_errs)):\n",
    "        fig = plt.figure(figcnt, figsize=(7.5, 5.625))\n",
    "        ax = plt.axes()\n",
    "        ax.plot(codoncounts, data[:,i])\n",
    "        plt.title(title_full + \", p_err: \" + str(p_errs[i]))\n",
    "        plt.xlabel('#Codons')\n",
    "        plt.ylabel('Probability')\n",
    "        plt.xlim([0, codoncounts[-1] + 1])\n",
    "        figerrs[i][0] = fig\n",
    "        if save: plt.savefig(\"output/perrfig/\" + str(statistic) + \"_line_perr\" + str(p_errs[i]) + \"_\" + title + \".pdf\")\n",
    "        if show: plt.show()\n",
    "        plt.close()\n",
    "        figcnt += 1\n",
    "        \n",
    "        fig = plt.figure(figcnt, figsize=(7.5, 5.625))\n",
    "        ax = plt.axes()\n",
    "        ax.bar(codoncounts, data[:,i],width=15/codoncounts[-1], color=[\"tab:blue\", \"tab:orange\", \"tab:green\", \"tab:red\"])\n",
    "        plt.title(title_full + \", p_err: \" + str(p_errs[i]))\n",
    "        plt.xlabel('#Codons')\n",
    "        plt.ylabel('Probability')\n",
    "        plt.xlim([0, codoncounts[-1] + 1])\n",
    "        figerrs[i][1] = fig\n",
    "        if save: plt.savefig(\"output/perrfig/\" + str(statistic) + \"_bar_perr\" + str(p_errs[i]) + \"_\" + title + \".pdf\")\n",
    "        if show: plt.show()\n",
    "        plt.close()\n",
    "        figcnt += 1\n",
    "    \n",
    "    return figheat, figcodonline, figperrline, figcods, figerrs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c1cf8de-3c9c-4e20-8185-d1d6fc8de166",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Error Modes (or Error Models)\n",
    "Defining the error modes\n",
    "An error mode is a tuple of two functions:\n",
    "- one for generating the error code and (modified character) from the character and error\n",
    "- another for getting a list of nucleotides that the error gets corrected to\n",
    "\n",
    "Current modes:\n",
    "- 01 : binary error code, 0 bit represents no error, 1 bit represents error (correct is other nucleotide than read RNA)\n",
    "- 100 : fully recoverable error code (stores sequence in error code), read RNA is uniform random per nucleotide\n",
    "- close : binary error code, 0 bit represents a nucleotide close to the correct one (or the correct one) was read, 1 bit represents a far away nucleotide to the correct one was read\n",
    "- real :  real-valued error code in range [0,1], value represents certainty of correctness (chosen arbitrarily in [0,1) in case of an error) where higher represents more certain of correctness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85077e34-aac2-44f6-8c6e-27d89041626b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# error bit 0 represents certain no error, error bit 1 represents certain error\n",
    "def errgen_01(char, err):\n",
    "    if err == 0:\n",
    "        return char, 0\n",
    "    return random.choice([x for x in corenucchars if not x == char]), 1\n",
    "\n",
    "def errcor_01(char, err):\n",
    "    if err == 0:\n",
    "        return [char]\n",
    "    \n",
    "    return [x for x in corenucchars if not x == char]\n",
    "\n",
    "em_01 = (errgen_01, errcor_01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b1c28e6-c4b0-4c26-890c-3d218a1035c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# error bit represents the correct nucleotide, \n",
    "def errgen_100(char, err):\n",
    "    return random.choice([x for x in corenucchars]), char\n",
    "\n",
    "def errcor_100(char, err):\n",
    "    return [err]    \n",
    "\n",
    "em_100 = (errgen_100, errcor_100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d9dbebc-4373-476c-ac3f-91d29069cf81",
   "metadata": {},
   "outputs": [],
   "source": [
    "# error bit 0 represents close nucleotide to original read, error bit 1 represents far nucleotide to original read\n",
    "def close_char(char):\n",
    "    close = set()\n",
    "    for p in closepairs:\n",
    "        if char in p:\n",
    "            close.update(p)\n",
    "    return close\n",
    "\n",
    "def errgen_close(char, err):\n",
    "    if err == 0:\n",
    "        return char, 0\n",
    "    \n",
    "    close = close_char(char)\n",
    "    mod_char = random.choice([x for x in corenucchars if not x == char])\n",
    "    \n",
    "    if mod_char in close:\n",
    "        return mod_char, 0\n",
    "    \n",
    "    return mod_char, 1\n",
    "    \n",
    "def errcor_close(char, err):\n",
    "    close = close_char(char)\n",
    "    if err == 0:\n",
    "        return close\n",
    "    return [x for x in corenucchars if not x in close]\n",
    "\n",
    "em_close = (errgen_close, errcor_close)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70e1e5e8-4202-494d-bb14-59b60eec6794",
   "metadata": {},
   "outputs": [],
   "source": [
    "# error bit (in [0,1]) represents how `confident' we are (higher is better)\n",
    "def errgen_real(char, err):\n",
    "    if err == 0:\n",
    "        return char, 1\n",
    "    return random.choice([x for x in corenucchars if not x == char]), random.random()\n",
    "\n",
    "def errcor_real(char, err):\n",
    "    aux = random.random()\n",
    "    if aux <= err:\n",
    "        return [char]\n",
    "    return [random.choice([x for x in corenucchars if not x == char])]\n",
    "\n",
    "em_real = (errgen_real, errcor_real)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5c79603-a5bf-4bde-97e2-0eee4426c518",
   "metadata": {},
   "source": [
    "## Finally, we call these functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc2da9ce-a20c-4c3b-9609-e6d64f8f49a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is how the RNA is generated for one experiment\n",
    "rna, err, mod_rna, err_code = gen_RNAerr(10, 0.3, em_close)\n",
    "print(rna)\n",
    "print(''.join(map(str,err)))\n",
    "print(mod_rna)\n",
    "print(''.join(map(str,err_code)))\n",
    "\n",
    "print(\"\")\n",
    "\n",
    "# This is how the RNA is generated for one experiment\n",
    "rna, err, mod_rna, err_code = gen_RNAerr(10, 0.3, em_close)\n",
    "print(rna)\n",
    "print(''.join(map(str,err)))\n",
    "print(mod_rna)\n",
    "print(''.join(map(str,err_code)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33ea9331-3da8-4406-a671-4e67fcbcf7b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is how an experiment is performed\n",
    "# it returns the:\n",
    "# - number of successes (full sequence successfully read\n",
    "# - number of errors successfully corrected\n",
    "# - number of successful codon reads given the number of errors per codon [0, 1, 2, 3 errors]\n",
    "# - number of total errors\n",
    "# - number of total codon reads given the number of errors per codon [0, 1, 2, 3 errors]\n",
    "do_experiment(10, 0.35, 5000, em_close) # (sequence length, error probability, number of trials, error mode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81af2d95-dce2-41f4-8eeb-f4f1b8cc8af4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is how we do multiple experiments (for several sequence lengths and error probabilities)\n",
    "# it returns a 2D list of the individual experiment results\n",
    "res = do_experiments([10,20], [0.01,0.1], 8196, em_real) # ([sequence lenghts], [error probabilities], nmber of trials, error mode)\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a68cf77-72da-4f6e-a9ac-fb443657f628",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# For saving: set save=True and make sure that in the directory of this notebook there exists the folders output>heatfig, output>codonfig, output>perrfig\n",
    "heat, codonlines, perrslines, codon, perrs = drawfigures_experiments([10,20], [0.01,0.1], 8196, res, statistic=0, title = \"Yadayada\", show=True, save=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2287faa7-0a41-4d50-8aad-8153f86811e4",
   "metadata": {},
   "source": [
    "## Number of Trials\n",
    "To determine a good number of trials to choose, we run a small set of experiments for several number of trials.\n",
    "We then pick a satisfactory one."
   ]
  },
  {
   "cell_type": "raw",
   "id": "1e73bc0a-15ae-45d7-9ed5-7ac624e3af38",
   "metadata": {},
   "source": [
    "codoncounts = [10,100]\n",
    "machineerrors = [0.001, 0.005, 0.01, 0.05, 0.1, 0.5]\n",
    "resses = []\n",
    "\n",
    "for i in range(8,15):\n",
    "    trials = 2**i\n",
    "    tstart = time.time()\n",
    "    resses.append(do_experiments(codoncounts, machineerrors, trials, em_01))\n",
    "    print(\"trials:\", trials, \"took\", time.time()-tstart)\n",
    "\n",
    "heats = []\n",
    "for i in range(8,15):\n",
    "    htmp = []\n",
    "    for j in range(0,7):\n",
    "        h, _, _, _, _ = drawfigures_experiments(codoncounts, machineerrors, 2**i, resses[i-8], statistic=j, title = str(2**i) + \" trials\", show=False, save=False)\n",
    "        htmp.append(h)\n",
    "    heats.append(htmp)\n",
    "    \n",
    "heats = np.array(heats, dtype=object)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "8c0ec693-85d0-4749-a8ce-8fbd40082dfc",
   "metadata": {},
   "source": [
    "statistic = 4\n",
    "for h in heats[:,statistic]:\n",
    "    display(h)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2c7719d-12c9-4f83-9850-ac36382a9c09",
   "metadata": {},
   "source": [
    "## Experiments\n",
    "We are now equipped to write the actual experiments.\n",
    "The number of trials is set to 2048 = 2^12."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bb7ac74-0a8e-4ac4-9c40-db527be0fd90",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this is how you save only one figure \n",
    "h = plt.figure(heat)\n",
    "#plt.savefig(\"tmp.pdf\") (UNCOMMENT THIS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70ea85c8-8576-4b77-b52e-3f426463468b",
   "metadata": {},
   "outputs": [],
   "source": [
    "codoncounts = [10, 50, 100, 500, 1000, 1500, 2000]\n",
    "machineerrors = [0.001, 0.005, 0.01, 0.05, 0.1, 0.5]\n",
    "errormode = em_01\n",
    "trials = 2048\n",
    "\n",
    "res = do_experiments(codoncounts, machineerrors, trials, errormode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6610e08-af97-4834-ba08-5e137b3d3366",
   "metadata": {},
   "outputs": [],
   "source": [
    "heat, codline, errline, figcods, figerrs = drawfigures_experiments(codoncounts, machineerrors, trials, res, statistic=0, title = \"Binary\", show=False, save=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf1fa7d4-178e-4582-bfc4-16715e9b73bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = plt.figure(heat)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
